app/change_set_persisters/hyrax/change_set_persister.rb

```diff
@@ -3,45 +3,107 @@
+      run_hook(:after_save, change_set: change_set)
+      query_service.find_by(id: resource.id)
     end
 
+    # Deletes the resource contained in the change_set from the adapters specified
+    # upon initialization of the ChangeSetPersister
+    # @param change_set Any change_set created by Hyrax::DynamicChangeSet
+    # @return [Boolean] true returned if the resource in the change_set was
+    #   succesfully deleted
+    # @return [Boolean] false returned if the resource in the change_set was not deleted
     def delete(change_set:)
-      catch(:abort) do
-        run_hook(:before_delete, change_set: change_set)
-        persister.delete(resource: change_set.resource)
-        run_hook(:after_delete, change_set: change_set)
-      end
+      run_hook(:before_delete, change_set: change_set)
+      persister.delete(resource: change_set.resource)
+      run_hook(:after_delete, change_set: change_set)
     end
 
+    # Calls save on all change_sets in the array passed in
+    # @param [Array] change_sets An array of any change_set created by Hyrax::DynamicChangeSet
+    # @return [Array] an array of the saved resources, such as GenericWork,
+    #  that is subclassed from Valkyrie::Resource
     def save_all(change_sets:)
+      objects = []
       change_sets.map do |change_set|
-        save(change_set: change_set)
+        objects << save(change_set: change_set)
       end
+      objects
     end
 
-    def buffer_into_index
-      metadata_adapter.persister.buffer_into_index do |buffered_adapter|
-        with(metadata_adapter: buffered_adapter) do |buffered_changeset_persister|
-          yield(buffered_changeset_persister)
-        end
+    # Calls delete on all change_sets in the arraypassed in
+    # @param [Array] change_sets An array of any change_set created by Hyrax::DynamicChangeSet
+    # @return [Array] an array of the resources that failed to delete, if empty
+    #  all resources were succesfully deleted
+    def delete_all(change_sets:)
+      change_sets.map do |change_set|
+        delete(change_set: change_set)
       end
     end
 
-    def with(metadata_adapter:)
-      yield self.class.new(metadata_adapter: metadata_adapter, storage_adapter: storage_adapter)
+    # Given a resource in a change_set, recursively deletes all members for the entire
+    # tree of members.  Will delete all members first and then the resource itself
+    # to avoid issues with removing the root of the tree first.  However all other
+    # resources are deleted as found by the query, so issues may arise from the order
+    # the resources were found in, so depending on the complexity of your member tree
+    # you may want to execerise caution in calling this method.
+    # @param [Array] change_sets An array of any change_set created by Hyrax::DynamicChangeSet
+    # @return [Array] an array of the resources that failed to delete, if empty
+    #  all resources were succesfully deleted
+    def delete_self_and_members(change_set:)
```

Comment:
Previously we were just deleting FileSets, but this is now deleting all members. Do we need to insure the members we are deleting are not members of other objects (not-deleted)?  This is one of the beneficial constraints on FileSets, they can only be a member of one object.

Reply:
My personal feeling is if you chose to call `delete_self_and_members` it's going to do that.  If you want to apply additional restrictions (type, membership to something else, etc) you can run `get_ids_of_self_and_all_members` (or some other query) and then prune out the `Valkyrie::ID`'s you don't want to delete before sending the array to `delete_all`.  

Before we didn't have delete recursion, so it was just `delete_all`, so for FileSets it would be a query for all members of the resource that are FileSets and pass the result to `delete_all`.  I'm definitely fine with putting in an issue for a recursive delete with some filtering support though.